### PR TITLE
[Graph Partition] add log for graph partition reasons and #partitions

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Previously, we log `skipping cudagraphs due to [xxx reasons]` when there are cudagraph-unsafe ops. With graph partition, we will split off these ops and cudagraph remaining parts. But the log message is also skipped.

In this PR, we add logs for graph partition reasons and the number of partitions to better understand the workload.

cc @mcarilli @ezyang @eellison @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: test/inductor/test_cudagraph_trees.py [hunk 1]
```diff
@@ -2710,6 +2710,20 @@ def f(x, y):
...
+        def test_graph_partition_log_message(self):
+            def foo(x, y):
+                return (x + 1, y + 2)
+
+            foo = torch.compile(foo, mode="reduce-overhead")
+
+            with capture_stderr() as captured_output:
+                foo(torch.ones([10], device="cuda"), torch.ones([20]))
+
+            FileCheck().check("cudagraph partition due to non gpu ops").check(
```
Comment (eellison) [inline test/inductor/test_cudagraph_trees.py]: nit: check_once ? seems like we had to avoid double warning

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -4220,27 +4221,44 @@ def can_buffer_be_removed_through_fusion(
...
             return any(self.should_partition(snode) for snode in node.snodes)
 
         if not node.is_gpu():
+            log_partition_reason("non gpu ops")
+
             return True
 
         if node.node is None:
+            log_partition_reason("None ops")
             return True
```
Comment (eellison) [inline torch/_inductor/scheduler.py]: Why do we skip when node.node is None ? when would this even happen?

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -4220,27 +4221,44 @@ def can_buffer_be_removed_through_fusion(
...
+        def noop_log(msg: str) -> None:
+            return
+
+        log_partition_reason = maybe_log_cudagraph_partition if should_log else noop_log
+
         if isinstance(node, FusedSchedulerNode):
             return any(self.should_partition(snode) for snode in node.snodes)
 
         if not node.is_gpu():
+            log_partition_reason("non gpu ops")
```
Comment (eellison) [inline torch/_inductor/scheduler.py]: add stack trace

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -4220,27 +4221,44 @@ def can_buffer_be_removed_through_fusion(
...
+
             return True
 
         if node.node is None:
+            log_partition_reason("None ops")
             return True
 
         if isinstance(node.node, ir.DeviceCopy):
+            log_partition_reason("DeviceCopy ops")
```
Comment (eellison) [inline torch/_inductor/scheduler.py]: stack trace

**Review by eellison — COMMENTED** [review]
looks good but these should all include stack trace as prior msgs do

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (BoyuanFeng) [reply torch/_inductor/scheduler.py]: Oh it was added before [FusedSchedulerNode is handled properly](https://github.com/pytorch/pytorch/pull/147038/commits/c70c46aa12464296a8eed0eaf0877ab1a3ef1cca), which have `node.node` as None and actual node in `node.snodes`.  For lint reasons, we can replace it with `assert node.node is not None` now.

File: torch/_inductor/utils.py [hunk 5]
```diff
@@ -3343,12 +3345,12 @@ def tabulate_2d(elements: Sequence[Sequence[T]], headers: Sequence[T]) -> str:
...
             widths[i] = max(widths[i], len(str(e)))
     lines = []
-    lines.append("|".join(f" {h:{w}} " for h, w in zip(headers, widths)))
+    lines.append("|".join(f" {h:OrderedSet([w])} " for h, w in zip(headers, widths)))
     #              widths          whitespace      horizontal separators
     total_width = sum(widths) + (len(widths) * 2) + (len(widths) - 1)
     lines.append("-" * total_width)
     for row in elements:
-        lines.append("|".join(f" {e:{w}} " for e, w in zip(row, widths)))
+        lines.append("|".join(f" {e:OrderedSet([w])} " for e, w in zip(row, widths)))
```
Comment (eellison) [inline torch/_inductor/utils.py]: i think you can noqa setlinter here, this is false positive. cc @rec  (who is already aware of this i believe)

File: torch/_inductor/utils.py [hunk 6]
```diff
@@ -3460,3 +3462,26 @@ def get_free_symbols(x: IterateExprs, unbacked_only: bool) -> OrderedSet[sympy.S
...
+def maybe_log_cudagraph_partition(
+    msg: str,
+    prefix: Optional[str] = "cudagraph partition due to ",
+    node: Optional[BaseSchedulerNode] = None,
+) -> None:
+    """
+    Cudagraph partition may lead to extra memory overhead so we
+    log partition reasons to help users understand the overhead.
+    """
+    if config.triton.cudagraphs:
```
Comment (eellison) [inline torch/_inductor/utils.py]: nit:: return early to reduce nesting. also would expect this always to be true if we're already here..

**Review by eellison — APPROVED** [review]
🚢
